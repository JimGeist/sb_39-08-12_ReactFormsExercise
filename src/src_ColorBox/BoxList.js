import React, { useState } from "react";
import NewBoxForm from "./NewBoxForm";
import Box from "./Box";

import './BoxList.css';

function BoxList() {

    const [boxes, setBoxes] = useState([]);

    const handleAddBox = (newBox) => {
        setBoxes(boxes => [...boxes, { ...newBox }]);
    }

    const handleDeleteBox = (id) => {
        setBoxes(boxes.filter(box => box.id !== id));
    }

    return (
        <div className="BoxList-Main">
            <NewBoxForm addBoxFx={handleAddBox} />
            <div>
                {
                    boxes.map(({ id, width, height, bgColor }) => (
                        <Box key={id} id={id} width={width} height={height} bgColor={bgColor} deleteFx={handleDeleteBox} />
                    ))
                }
            </div>

        </div>
    );

}

export default BoxList;
