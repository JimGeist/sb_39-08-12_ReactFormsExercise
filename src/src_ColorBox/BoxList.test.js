import React from "react";
import { render, fireEvent } from "@testing-library/react";
import BoxList from "./BoxList";

// Smoke Test -- does BoxList render?
it("renders the BoxList component without crashing", () => {
    render(<BoxList />);
});

// BoxList snapshot tests
it("BoxList component matches snapshot", () => {
    const { asFragment } = render(<BoxList />);
    expect(asFragment()).toMatchSnapshot();
});

it("has a width, height, and color labels and an Add Box button", function () {
    const { getByLabelText, getByText } = render(<BoxList />);

    // expect 'Box Width:', 'Box Height:', and 'Box Color:' labels
    expect(getByLabelText("Box Width:")).toBeInTheDocument();
    expect(getByLabelText("Box Height:")).toBeInTheDocument();
    expect(getByLabelText("Box Color:")).toBeInTheDocument();

    // expect 'Add Box' button
    expect(getByText("Add Box")).toBeInTheDocument();

});

it("can add a new box and delete it", function () {
    const { queryByTestId, getByLabelText, queryByText } = render(<BoxList />);

    // no boxes exist
    expect(queryByTestId("^Color Box^")).not.toBeInTheDocument();

    const boxWidth = getByLabelText("Box Width:");
    const boxHeight = getByLabelText("Box Height:");
    const boxColor = getByLabelText("Box Color:");
    const boxAddBtn = queryByText("Add Box")

    // fill out the form
    fireEvent.change(boxWidth, { target: { value: 100 } });
    fireEvent.change(boxHeight, { target: { value: 100 } });
    fireEvent.change(boxColor, { target: { value: "chartreuse" } });
    fireEvent.click(boxAddBtn);

    // box now exists
    expect(queryByTestId("^Color Box^")).toBeInTheDocument();

    const boxDeleteBtn = queryByText("X");
    fireEvent.click(boxDeleteBtn);

    // box no longer exists
    expect(queryByTestId("^Color Box^")).not.toBeInTheDocument();


});
