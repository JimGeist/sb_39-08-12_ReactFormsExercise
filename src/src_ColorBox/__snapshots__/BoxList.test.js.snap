// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BoxList component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="BoxList-Main"
  >
    <form
      class="NewBoxForm"
    >
      <label
        for="width"
      >
        Box Width:
      </label>
      <input
        class="NewBoxForm-InpNbr"
        id="width"
        name="width"
        type="number"
        value=""
      />
      <label
        for="height"
      >
        Box Height:
      </label>
      <input
        class="NewBoxForm-InpNbr"
        id="height"
        name="height"
        type="number"
        value=""
      />
      <label
        for="bgColor"
      >
        Box Color:
      </label>
      <input
        id="bgColor"
        name="bgColor"
        type="text"
        value=""
      />
      <button
        class="NewBoxForm-BtnAdd"
      >
        Add Box
      </button>
    </form>
    <div />
  </div>
</DocumentFragment>
`;
